python/paddle/fluid/layers/tensor.py

```diff
@@ -378,6 +379,97 @@ def concat(input, axis=0, name=None):
+
+    Args:
+        input(list|tuple): ``input`` is a Tensor list or Tensor tuple which is with data type bool,
+            float32, float64, int32, int64. All the Tensors in ``input`` must have same data type.
+        name (str, optional): The default value is None. Normally there is no need for user to set this property. 
+            For more information, please refer to :ref:`api_guide_Name`.
+
+    Returns:
+        list(Tensor): The list of broadcasted tensors following the same order as ``input``.
+
+    Examples:
+        .. code-block:: python
+            import paddle
+            import numpy as np
+            
+            x1 = paddle.to_tensor(np.random.random([1, 2, 3, 4]).astype('float32'))
+            x2 = paddle.to_tensor(np.random.random([1, 2, 1, 4]).astype('float32'))
+            x3 = paddle.to_tensor(np.random.random([1, 1, 3, 1]).astype('float32'))
+            out1, out2, out3 = paddle.broadcast_tensors(x=[x1, x2, x3])
+            # out1, out2, out3: tensors broadcasted from x1, x2, x3 with shape [1,2,3,4]
+    """
+    check_type(input, 'input', (list, tuple), 'broadcast_tensors')
+    num_inputs = len(input)
+    if num_inputs < 2:
+        raise TypeError(
+            "number of input tensors to broadcast_tensors should be no less than 2"
+        )
+
+    # Check input types
+    for id, x in enumerate(input):
+        check_variable_and_dtype(
+            x, 'input[' + str(id) + ']',
+            ['bool', 'float32', 'float64', 'int32', 'int64'],
+            'broadcast_tensors')
+        if x.dtype != input[0].dtype:
+            raise TypeError(
+                "All the Tensors in the input must have the same data type.")
+
+    # Check bcast semantics
+    output_shape_r_last_tensor_index = []
+    output_shape_r = []
+
+    # Use while loop due to weird behaviour of "range()"
+    j = 0
+    while j < len(input):
+        tensor = input[j]
+        shape = list(reversed(tensor.shape))
+
+        i = 0
+        while i < len(shape):
+            if len(output_shape_r) <= i:
+                output_shape_r.append(shape[i])
+                output_shape_r_last_tensor_index.append(j)
+            else:
+                invalid = (output_shape_r[i] != shape[i] and
+                           output_shape_r[i] != 1 and shape[i] != 1)
+                if invalid:
+                    last_index = output_shape_r_last_tensor_index[i]
+                    raise TypeError(
+                        "Input tensors to broadcast_tensors does not follow bcast semantics"
+                        f"Tensor {last_index} conflicts with Tensor {j} in reversed dimension {i}"
+                    )
+                if output_shape_r[i] <= shape[i]:
+                    output_shape_r[i] = shape[i]
+                    output_shape_r_last_tensor_index[i] = j
+            i += 1  # while i < len(shape)
+        j += 1  # while j < len(input)
+
+    if in_dygraph_mode():
```

Comment:
the performance of this writing method is bad, put the dygraph branch before checks? or put the necessary checking in C++ for dygraph?

Reply:
Moved dygraph branch to the very beginning of the API function
Added bcast semantics check at C++ level, inside Op::InferShape().